README
NAME
    App::Twirc::Manual - User guide and reference for Twirc

DESCRIPTION
    "Twirc" is a twitter client. More precisely, it is an IRC/Twitter
    gateway that makes your favorite IRC client a twitter client.

FEATURES
    Follow friends timelines
    Receive replies from friends and non-friends
    Post status updates
    Send and receive direct messages
    Follow, un-follow, block, unblock, and use most other twitter commands

INSTALLATION
    You can install "twirc" just as you would any other CPAN distribution:

        cpan POE::Component::Server::Twirc

    Or, you can download and unpack "twirc", then run it from a directory of
    your choice without running "make install". That's the way the author
    runs it. If you choose this option, run "make" to install "twirc"'s
    dependencies.

CONTRIBUTING AND BUILDING
    If you'd like to contribute to "twirc" or just want to build the
    distribution for some other reason, you must have the following Perl
    modules:

    Module::Install
    Module::Install::Contributors
    Module::Install::Repository
    Module::Install::AutoManifest

CONFIGURATION
    "Twirc" uses Config::Any, so you can configure "twirc" using XML, YAML,
    JSON, Apache-style configuration, Windows INI file format, or even Perl
    code.

    A configuration file is not necessary, but is recommended.

    Here's an example configuration in YAML:

        state_file: twirc.state
        log_level: INFO

  CONFIGURATION OPTIONS
    irc_server_name
        The name of the IRC server. Defaults to "twitter.irc". Every IRC
        server has a name. The IRC server included with "twirc" isn't
        intended to be accessed publicly. It is for your own personal use.
        So, the name is not significant.

    irc_server_port
        The port number the IRC server binds to. Defaults to 6667.

    irc_server_bindaddr
        The local address to bind to. Defaults to all 127.0.0.1 to prevent
        others from connecting to your "twirc" IRC server. If you run
        "twirc" on a different system than your IRC client, you will need to
        provide this configuration option with a suitable address.

    irc_mask
        The IRC user/host mask used to restrict connecting users. Defaults
        to "*@127.0.0.1". If you run "twirc" on a different system than your
        IRC client, you will need to provide this configuration option with
        a suitable mask.

    irc_password
        Password used to authenticate to the IRC server. If you don't
        provide this option, no password will be required. It adds a bit of
        security. You may want to set this option if other users have access
        to your system.

    irc_botname
        The name of the channel operator bot. Defaults to "tweeter". Select
        a name that does not conflict with friends, followers, or your own
        IRC nick.

        When running "twirc", you interact with a bot in the channel. The
        bot carries out commands on your behalf and provides feedback,
        particularly when there are errors.

    irc_botircname
        Text to be used as the channel operator bot's IRC full name.
        Defaults to "Your Friendly Twitter Agent". This is the name that
        will appear in response to an IRC "/whois" command.

    irc_channel
        The name of the channel where your twitter friends' timelines
        appear. This is the channel where most of your interaction with
        "twirc" occurs. It defaults to &twitter. The IRC convention for
        channels names is channels local to a single server begin with "&".
        Network channels begin with "#". You can use either to name, however
        "&" is more appropriate.

    twitter_alias
        An alias to use for displaying incoming status updates from the
        owning user. This is necessary if the user's IRC nickname and
        Twitter screen name are the same. Defaults to "me".

        With the default value "me", when "twirc" reads a status message in
        your timeline from your Twitter screen name, it will use "me" in
        place of your Twitter screen name in the channel.

    selection_count
        How many status messages to display for selection when favoriting,
        replying, or retweeting. Defaults to 3.

    truncate_to
        When displaying a list tweets for selection, for example, in
        response to the "favorite" command, they will be truncated to this
        length to avoid cluttering the screen with long messages that wrap.
        Defaults to 60.

    log_channel
        If specified, twirc will post log messages to this channel. If you
        set this option to &log, then you can join the &log channel and see
        the copious debug messages that "twirc" generates. This may be
        useful for trouble shooting or problem reporting.

    log_level
        Twirc supports "log_level" values OFF, FATAL, ERROR, WARN, INFO,
        DEBUG, and TRACE. The default is WARN.

    state_file
        File used to store state information between sessions, including
        Twitter OAuth access tokens, friends, and followers_ids.

    twitter_args
        Use this to pass additional parameters to Net::Twitter's
        constructor. For example, this version of Twirc uses SSL connections
        to the Twitter API. If you're using a service that requires non-SSL
        connections, add this to your configuration file:

            twitter_args:
                ssl: 0

USING
    To use "twirc" you first need to start the server:

        bin/twirc -b --state_file=twirc.state

    The "-b" option runs "twirc" in the background. Drop the "-b" to see log
    messages to STDERR. (The author runs twirc and his irc client in screen,
    <http://www.gnu.org/software/screen/>, to monitor log messages to
    STDERR.)

    Next, connect to the server from your IRC client. I use "irssi"
    (<http://www.irssi.org>) and my examples will use "irssi" commands:

        /connect localhost

    On connection, "twirc" will automatically join you to the configured
    channel. The default &twitter will be assumed, here.

    Your friends will be automatically joined to the channel. Friends who
    are also followers are given voice as a visual clue. In "irssi" they
    appear with plus sign (+) in front of their names.

    To post a new status update, use the "post" command:

        post My first tweet from twirc!

    In general, you enter a command followed by its arguments, if any, as a
    public message in the channel. There's a handy exception to that rule
    for sending replies. An entry that begins with a nick name, followed by
    a colon is treated as a reply. E.g.:

        twirc: you make twitter usable!

    Is a shortcut for:

        post @twirc you make twitter usable!

    "twirc" uses the Twitter User Streams API to receive updates in
    real-time.

    Use IRC private messaging to send direct messages. In "irssi":

        /msg friend Watch out, I'm right behind you!

    The "twirc" server stops when you disconnect. This isn't normal IRC
    behavior, but "twirc" isn't a normal IRC server. Its only purpose is to
    interface with Twitter on your behalf and server no useful purpose when
    you're not connected.

COMMAND REFERENCE
    post status
        Post a status update. E.g.,

            post Now cooking tweets with twirc!

    follow twitter_screen_name
        Follow a new Twitter user. This creates a friend relationship and
        adds the friend to the channel.

    unfollow twitter_screen_name
        Stop following a Twitter friend. This destroys the friend
        relationship and removes the friend from the channel.

    block twitter_screen_name
        Blocks the Twitter user from receiving your Twitter feed.

    unblock twitter_screen_name
        Stop blocking a Twitter user.

    whois twitter_user
        Displays information about Twitter user. "twitter_user" can be
        either a screen name or email address.

    notify on|off twitter_screen_name...
        Turns device notifications on or off for the list of one or more
        Twitter friends. The list is space separated.

    retweets on|off twitter_screen_name...
        Turns retweet display on your timeline on or off for the list of one
        or more Twitter friends. The list is space separated.

    favorite friend [ count ]
        Mark a friend's tweet as a favorite. Optionally, specify the number
        of tweets to display for selection with "count". ("count" defaults
        to 3. The default can be changed with the "favorites_count" option.)

    rate_limit_status
        Displays information about the remaining number of API requests
        available in the current hour. The "rate_limit_status" command does
        not count against the limit, itself.

    help
        Display a simple help message listing the available command names.

TIPS AND TRICKS
  Ignoring
    If you're following a particularly noisy friend, you can of course
    "unfollow" her. Alternatively, you can use your IRC client's features to
    ignore the user. In "irssi":

        /ignore LoudMouth ALL
        /ignore -except -pattern @YourName LoudMouth ALL

    Now, you won't hear from LoudMouth unless she @replies you, and she
    won't know you're ignoring her (unless she reads this tip!).

  Multiple accounts
    Although "twirc" doesn't currently support multiple accounts, there's
    nothing stopping you from running multiple instances of "twirc", one for
    each account.

    Here's how I do it for accounts "semifor" (my personal account) and
    "twirc" ("twirc"'s feed for update notices, etc.).

    First, the pertinent sections of the configuration files (in YAML
    format).

        # File: semifor.yml
        irc_server_port: 6667
        irc_password: secret
        irc_channel: '&twitter'

        # File: twirc.yml
        irc_server_port: 6668
        irc_password: secret
        irc_channel: '&twirc'

    Next, start an instance for each account:

        bin/twirc -c semifor.yml -b
        bin/twirc -c twirc.yml -b

    In your IRC client, connect to both instances ("irssi" here):

        /connect localhost 6667 secret semifor
        /connect localhost 6668 secret twirc

    Now you've got 2 channels, one for each account---in my case, &twitter
    for "semifor" and &twirc for "twirc".

  PLUGINS
    Plugins are modules that are optionally included by specifying them in
    user configuration file in the "plugins" section. Some of the plugins
    included in the distribution are:

    SquashWhiteSpace
        Squashes whitespace in tweets to a single space. See
        App::Twirc::Plugin::SquashWhiteSpace.

    BangCommands
        Commands are prefixed with a bang (i.e., exclamation point "!").
        Text entered without a bang prefix is posted as a status update. See
        App::Twirc::Plugin::BangCommands.

    SecondaryAccount
        Cross-posts status updates to one or more secondary Twitter (or
        Twitter API compatible, like Identi.ca) accounts. See
        App::Twirc::Plugin::SecondaryAccount.

SUPPORT
    "Twirc" is free open source software with no warranty of any kind. That
    said, it's used by some competent perl coders who may be able to help if
    you have trouble. Try the "#net-twitter" channel at "irc.perl.org".

    The code repository with the development branch is located at
    <http://github.com/semifor/twirc>. New features, and bug fixes appear
    there before they hit CPAN, so check the commit log there to see if a
    problem you've found has been addressed. And feel free to use the
    development branch.

CONTRIBUTORS
    Peter Roberts <me+dev@peter-r.co.uk>

AUTHOR
    Marc Mims <marc@questright.com>

    I'm "semifor" on twitter and IRC.



Changes
    - Reply to original status when replying to retweets

0.20 2014-03-30
    - Added PkgVersion so dzil will version the packages!

0.19 2014-03-30

0.18-TRIAL  2014-03-26
    - Major overhaul
      - fixed circular references
      - full, non-blocking IO for Twitter API callsv
      - Many fixes by Peter Roberts (Thank you!)

0.17 2014-01-18
    - Enable SSL for Twitter API calls

0.16 2013-06-13
    - initial Twitter API v1.1 support (BinGOs)

0.15 2012-05-09
    - Fixed: exit server on disconnect
    - Bumped requried version of MooseX::POE (resovles failed CPAN Testers report)

0.14 2012-05-08
    - use Twitter User Streams API instead of polling
    - use PIN based Twitter OAuth authentication (Twitter dropped our xAuth access)
    - simplified configuration (config and state files are now optional)
    - add proper POE fork code (Apocalypse)

0.13 2010-12-10
    - added command `rt` alias for `retweet`
    - treat Twitter status IDs as strings (they now exceed Int capacity on 32-bit systems)
    - fixed: irc_server_bindaddr type constraint
    - updated dependencies
    - filter bogus results from user_timeline (twitter bug)

0.12 2010-08-23
    - added dependency on MooseX::AttributeHelpers

0.11 2010-08-21
    - Use OAuth
    - Show full text of retweeted messages
    - usage messages for retry/retweet commands
    - Added search plugin (Adam Prime)
    - Let plugins specify traits for Net::Twitter (Adam Prime)
    - Added reply command (sets in_relpy_to_status_id)
    - Use home_timeline instead of friends_timeline
    - Added retweet command for official Twitter retweets
    - Added report_spam command

0.10 2009-11-02
    - Replaced Text::Truncate with String::Truncate

0.09 2009-09-25
    - Replaced "page" parameter usage with "cursor" (friends/followers)
    - Fixed missing sprintf in whois handler

0.08 2009-08-17
    - Fixed: s/irc_alias/twitter_alias/
    - Fixed: dereferencing an undefiend user object

0.07 2009-07-21
    - Use Net::Twitter 3.x with native error handling and object inflation
    - Refactored Net::Twitter calls to workaround twitterapi and NT bugs
    - Fix: allow follow of a non-friend channel user
    - Nicks change when users' screen names change
    - Support for plugins
    - SquashWhiteSpace plugin
    - BangCommands plugin
    - SecondaryAccount plugin
    - Include META.yml in MANIFEST

0.06  2009-01-18
    - work around back compat bugs in Net::Twitter 2.00 and 2.01
    - fixed: display of incoming messages stopped after unfollow
    - trim trailing space from user input

0.05  2009-01-14
    - fixed topic setting: not on @replies, please
    - made verbose_refresh message optional
    - fixed: topic set twice (on post and on poll)

0.04  2009-01-13
    - added App::Twirc::Manual
    - added irc_server_bindaddr option (Sophie Hamilton)
    - added direct message checking
    - added rate_limit_status command
    - additional command feedback (rjbs, marcus)
    - Logging to channel &twirc-log
    - Twitter assigned source ID

0.03  2009-01-05
    - Released to CPAN
    - Renamed Net::Twitter::IRC to POE::Component::Server::Twirc
    - added backgrounding (--background)
    - temporarily work around JSON::Any fault on HTTP status 500
    - No logging by default; config option log_level for noise
    - Renamed twitirc.pl to twirc
    - Added check_replies and @replies merging
    - Use useragent_class in Net::Twitter 1.22
    - Added echo_posts option, defaults to 0
    - Net::Twitter 1.21 works around Twitter API fail on friendships/exists
    - error on messages longer than 140 characters
    - easy replies: "nick: ..." becomes "post @nick ..."

0.02_4  2008-12-16
	- disabled call to relationship_exists until fixed by Twitter or
	  Net::Twitter workaronud patch is applied

0.02_3  2008-12-16
    - give voice to existing followers on follow
    - Net::Twitter 1.20 fixes >100 followers

0.02_2  2008-12-15
    - fixed page param handling

0.02_1  2008-12-14
    - added support for >100 friends
    - added support for >100 followers (pending Net::Twitter patch)
    - added implementation for IRC mask/password auth

0.02  2008-12-14
    - decode HTML entities
    - Use LWP::UserAgent::POE for non-blocking IO
    - Help added (marcusramberg)
    - dependencies (marcusramberg)

0.01  2008-12-06
    - initial dev release


Makefile.PL
use inc::Module::Install;
use strict;
use warnings;

name     'POE-Component-Server-Twirc';
perl_version '5.008001';
all_from 'lib/POE/Component/Server/Twirc.pm';

install_script  'bin/twirc';

# Needed to run this script
configure_requires 'Module::Install::Repository';
configure_requires 'Module::Install::AutoManifest';
configure_requires 'Module::Install::Contributors';

requires 'AnyEvent::Twitter::Stream'   => '0.23'; # APIv1.1
requires 'Config::Any';
requires 'Encode';
requires 'FindBin';
requires 'HTML::Entities';
requires 'JSON::Any';
requires 'JSON::MaybeXS';
requires 'Log::Log4perl';
requires 'LWP::UserAgent::POE'         => '0.02';
requires 'Moose';
requires 'MooseX::Getopt'              => '0.15';
requires 'MooseX::Log::Log4perl::Easy';
requires 'MooseX::POE'                 => 0.215;
requires 'MooseX::SimpleConfig';
requires 'MooseX::Storage';
requires 'Net::Twitter'                => '4.00006'; # xauth + APIv1.1
requires 'Path::Class::File';
requires 'POE::Component::Server::IRC' => 0.02005;
requires 'POE::Component::TSTP';
requires 'Proc::Daemon';
requires 'Regexp::Common::URI';
requires 'Scalar::Util';
requires 'String::Truncate';
requires 'Try::Tiny';

if ( $Module::Install::Author ) {
    system 'pod2text lib/App/Twirc/Manual.pod > README';
}

contributors 'Adam Prime <adam.prime@utoronto.ca>';
contributors 'Peter Roberts <me+dev@peter-r.co.uk>';

auto_set_repository;
homepage   'https://github.com/semifor/twirc';
bugtracker 'https://github.com/semifor/twirc/issues';

auto_manifest;
auto_install;

WriteAll;
